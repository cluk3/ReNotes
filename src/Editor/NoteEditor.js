import React, { Component, Suspense } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import {
  updateEditorState,
  createNewNote,
  setActiveNote
} from 'Notes/modules/notes';
import styled from 'styled-components';
import { ENTITIES } from 'constants.js';
import format from 'date-fns/format';
// import ReactQuill from 'react-quill';
import 'react-quill/dist/quill.snow.css';
import './customQuill.css';

const ReactQuill = React.lazy(() => import('react-quill'));

const NoteEditorContainer = styled.div`
  height: 100%;
  max-height: 100%;
  overflow: auto;
  padding: 0.5em 1em 1em 1em;
  border-left: #dedede solid 1px;
`;

const LastModified = styled.span`
  margin-bottom: 1em;
  display: block;
  text-align: center;
  color: #555;
  font-size: 14px;
  font-weight: 300;
`;

export class NoteEditor extends Component {
  static propTypes = {
    editorState: PropTypes.object,
    activeNote: PropTypes.string,
    parentFolderName: PropTypes.string,
    setActiveNote: PropTypes.func.isRequired,
    createNewNote: PropTypes.func.isRequired,
    updateEditorState: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      loaded: false
    };
    this.onChange = (content, delta, source, editor) => {
      this.props.updateEditorState(
        editor.getContents(),
        editor.getText(),
        this.props.activeNote
      );
    };
  }

  componentDidUpdate(prevProps) {
    const {
      isEditorFocused,
      createNewNote,
      parentFolderName,
      editorState
    } = this.props;
    const editorGainedFocus = !prevProps.isEditorFocused && isEditorFocused;

    if (!editorState && editorGainedFocus) {
      return createNewNote(parentFolderName);
    }
  }

  componentDidMount() {
    window.addEventListener('DOMContentLoaded', () => {
      this.setState({
        loaded: true
      });
    });
  }

  render() {
    return (
      <NoteEditorContainer>
        <LastModified>
          {this.props.lastModified &&
            format(this.props.lastModified, 'D MMMM YYYY [at] h:mm A')}
        </LastModified>
        <Suspense fallback={<div>Loading...</div>}>
          {this.state.loaded &&
            this.props.editorState && (
              <ReactQuill
                theme="snow"
                value={this.props.editorState.contents}
                onChange={this.onChange}
                modules={{ toolbar: false }}
              />
            )}
        </Suspense>
      </NoteEditorContainer>
    );
  }
}

function mapStateToProps({ notes, folders, focusedElement }) {
  const { activeNote } = notes;
  const note = notes.byId[notes.activeNote];
  return {
    editorState: activeNote ? note.editorState : null,
    activeNote,
    lastModified: note && note.lastModified,
    parentFolderName: folders.activeFolder,
    isEditorFocused: focusedElement.elementType === ENTITIES.EDITOR
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      createNewNote,
      updateEditorState,
      setActiveNote
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(NoteEditor);
