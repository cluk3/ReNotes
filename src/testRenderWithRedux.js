import React from 'react';
import { createStore } from 'redux';
import { Provider, connect } from 'react-redux';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { initialState as notesInitialState } from './Notes/modules/notes';
import { initialState as foldersInitialState } from './Folders/modules/folders';
import { initialState as focusedElementInitialState } from './containers/Focusable/stateManager';

const reducer = state => state;
const defaultInitialState = {
  folders: foldersInitialState,
  notes: notesInitialState,
  focusedElement: focusedElementInitialState
};

export function renderWithRedux(
  ui,
  {
    initialState = defaultInitialState,
    store = createStore(reducer, initialState)
  } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    // adding `store` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    store
  };
}
