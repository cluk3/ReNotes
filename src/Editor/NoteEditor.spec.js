import React from "react";
import NoteEditor from "./NoteEditor";
import { renderWithRedux } from '../testRenderWithRedux'
import 'jest-styled-components';

describe("NoteEditor", () => {

  it('works', () => {
    const { container } = renderWithRedux(<NoteEditor />);
    expect(container.firstChild).toMatchSnapshot()
  })

  // it("should be initialized with no text if it's a new note", () => {});
  // it("should be initialized with no text if folder is empty", () => {});
  // it("should be initialized text from the active note", () => {});

  // it("should debounce the update to redux", () => {});

  // it("should save the text in the state", () => {});

  // it("should create a new note if clicked when folder is empty");
  //   describe("when Note is clicked", () => {
  //     it("should set item to delete", () => {

  //     });

  //     it("should set active note", () => {
  //     });
  //   });

  //   describe("when New Note button is clicked", () => {
  //     it("should create a new note", () => {
  //     });

  //     it("should set item to delete", () => {

  //     });

  //     it("should set active note", () => {

  //     });
  //   });
});
