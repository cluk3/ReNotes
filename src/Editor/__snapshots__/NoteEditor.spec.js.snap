// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteEditor works 1`] = `
.c0 {
  height: 100%;
  max-height: 100%;
  overflow: auto;
  padding: 0.5em 1em 1em 1em;
  border-left: #dedede solid 1px;
}

.c1 {
  margin-bottom: 1em;
  display: block;
  text-align: center;
  color: #555;
  font-size: 14px;
  font-weight: 300;
}

<div
  class="c0"
>
  <span
    class="c1"
  >
    23 June 2019 at 5:07 PM
  </span>
</div>
`;
